File: A4/triangles.py
import sys
from math import *


def read():
    rows_str = sys.stdin.readline()
    rows_num = int(rows_str)
    triangles = []
    for i in range(rows_num):
        line = sys.stdin.readline().strip("\n")
        line1 = line.split(' ')
        sideA = int(line1[0])
        sideB = int(line1[1])
        sideC = int(line1[2])
        shade = float(line1[3])
        # list.append((int(sideA), int(sideB), int(sideC), float(shade)))
        triangles.append(Triangle(sideA, sideB, sideC, shade))
        triangles.sort()
    return triangles

# let m(i) be max possible chain with ith triangle at the end
# m(i)= m(j) + 1, if j.shade < i.shade, one side matches
# if no such j then m(i) = 1


class Triangle:
    def __init__(self, sideA, sideB, sideC, shade):
        self.sides = (sideA, sideB, sideC)
        self.shade = shade

    def __lt__(self, other):
        return self.shade < other.shade

    def __repr__(self):
        return f"({self.sides} {self.shade})"


def _matrix(triangles):
    n = len(triangles)
    # print(n)
    matrix = [[1 for i in range(3)] for j in range(n)]

    for i in range(1, n):
        for j in range(3):
            maxx = 1
            for k in range(i):
                if triangles[i].shade <= triangles[k].shade:
                    continue
                for l in range(3):
                    if triangles[k].sides[(l+1) % 3] == triangles[i].sides[j] or\
                            triangles[k].sides[(l+2) % 3] == triangles[i].sides[j]:
                        maxx = max(maxx, matrix[k][l] + 1)
            matrix[i][j] = maxx
    # print(matrix)

    maxx = -1
    for i in range(n):
        for j in range(3):
            maxx = max(maxx, matrix[i][j])
    print(maxx)


if __name__ == "__main__":
    triangles = read()
    _matrix(triangles)

File: A4/triangles.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class triangles {

    public static void main(String[] args) throws IOException {
        InputStreamReader in = new InputStreamReader(System.in);
        BufferedReader input = new BufferedReader(in);

        int n = Integer.parseInt(input.readLine());
        Triangle[] triangleList = new Triangle[n];

        if (n <= 0) {
            return;
        }

        for (int i = 0; i < n; i++) {
            String line = input.readLine();
            String[] splitLine = line.split(" ");
            triangleList[i] = new Triangle(
                    Integer.parseInt(splitLine[0]),
                    Integer.parseInt(splitLine[1]),
                    Integer.parseInt(splitLine[2]),
                    Double.parseDouble(splitLine[3])
            );
        }

        Arrays.sort(triangleList);

        // mem[t][s] represents the length of the longest chain formed that ends with triangle t,
        // where t is joined to the chain using side s.
        int[][] mem = new int[n][3];

        // base cases
        for (int s = 0; s < 3; s++) {
            mem[0][s] = 1;
        }

        for (int t = 1; t < n; t++) {
            for (int s = 0; s < 3; s++) {
                int max = 1;
                for (int prevTriangle = 0; prevTriangle < t; prevTriangle++) {
                    if (triangleList[prevTriangle].shade >= triangleList[t].shade) {
                        continue;
                    }
                    for (int joinedEdge = 0; joinedEdge < 3; joinedEdge++) {
                        if (triangleList[prevTriangle].sides[(joinedEdge + 1) % 3] == triangleList[t].sides[s]
                                || triangleList[prevTriangle].sides[(joinedEdge + 2) % 3] == triangleList[t].sides[s]) {
                            max = Math.max(max, mem[prevTriangle][joinedEdge] + 1);
                        }
                    }
                }
                mem[t][s] = max;
            }
        }

        int max = 0;
        for (int t = 0; t < n; t++) {
            for (int s = 0; s < 3; s++) {
                max = Math.max(max, mem[t][s]);
            }
        }
        System.out.println(max);

    }
}

class Triangle implements Comparable<Triangle> {
    int[] sides = new int[3];
    double shade;

    public Triangle(int side0, int side1, int side2, double shade) {
        this.sides[0] = side0;
        this.sides[1] = side1;
        this.sides[2] = side2;
        this.shade = shade;
    }

    @Override
    public int compareTo(Triangle o) {
        return Double.compare(this.shade, o.shade);
    }

}